control-plane-operator/releaseinfo/pod_provider.go

```diff
@@ -34,10 +36,15 @@ func (p *PodProvider) Lookup(ctx context.Context, image string) (releaseImage *R
 			RestartPolicy: corev1.RestartPolicyNever,
 			Containers: []corev1.Container{
 				{
-					Name:    "lookup",
+					Name:    "read-image-references",
 					Image:   image,
 					Command: []string{"/usr/bin/cat", "/release-manifests/image-references"},
 				},
+				{
+					Name:    "read-coreos-metadata",
+					Image:   image,
+					Command: []string{"/usr/bin/cat", "/release-manifests/0000_50_installer_coreos-bootimages.yaml"},
```

Comment:
is this invoked in a shell where we can use wildcards?  And maybe regardless, might be worth PRing a comment to the installer source like "renaming this manifest will break a hard-coded path in the HyperShift contro-plane operator: $LINK".  Or can we make this lookup robust to installer-side renames by [loading manifests][1] and then iterating over them to find the one we want?

[1]: https://github.com/openshift/library-go/blob/e555322cb70844f90b003cbdfe7ce002d7c61810/pkg/manifest/manifest.go#L69-L71

Reply:
Couple of things we use which definitely need more thought:

- /release-manifests/image-references 
- /release-manifests/0000_50_installer_coreos-bootimages.yaml

In a given release payload, some questions about those files:

* Are they considered part of the payload "contract" somehow, e.g. by nature of having a well-known location?
* Are their contents guaranteed to be serializations of some public API types? For example, image-references is a serialized `ImageStream`, but bootimages is a serialized `ConfigMap` which embeds a JSON document whose schema I'm not aware of

Reply:
cc @derekwaynecarr

Reply:
`/release-manifests/image-references` is payload contract, e.g. [here][1] and [here][2] is the CVO requiring that file to exist.  The CVO requires manifests to exist in `release-manifests/`, uses their filenames for ordering, but otherwise isn't opinionated about what they are called, and does not require any manifest-filename consistency between releases.

It's possible that the installer might drop their boot-image ConfigMap, but seems unlikely, and even less likely if you let them know that you're consuming it.

[1]: https://github.com/openshift/cluster-version-operator/blob/77a24f77b8f6e9975b5f41a787d6368ff609e391/pkg/payload/payload.go#L97
[2]: https://github.com/openshift/cluster-version-operator/blob/77a24f77b8f6e9975b5f41a787d6368ff609e391/pkg/payload/payload.go#L244-L245

Reply:
I wonder if it makes sense for the OS metadata bits to become part of the contract? Otherwise it's not clear to me how providers would piece together the compatible payload/AMI (for example) versions from the outside in order to specify the AMI explicitly on the HostedCluster. @derekwaynecarr, thoughts here?

Reply:
where in the installer code is this name coming from **0000_50_installer_coreos-bootimages**?

Reply:
Sorry if this is a silly question, but is it possible for the ImageStream and CoreOS info to be embedded into the container metadata, serialized as values in labels or something?

Reply:
I'm going to lift this discussion out into our next meeting to get on the same page with the team before bringing the topic to other groups, I think it's a bigger subject than can be addressed in this PR

Reply:
> Otherwise it's not clear to me how providers would piece together the compatible payload/AMI (for example) versions from the outside...

Installer folks have traditionally aregued that their AMI defaults are internal and not something they conveniently expose to others.  Currently the supported path for user-provided node's is to look up the AMI in an openshift-docs table.  But yeah, personally I don't see a reason this can't be a label on the machine-os-content image or whatever.

> where in the installer code is this name coming from 0000_50_installer_coreos-bootimages?

The installer [puts it in `manifests/coreos-bootimages.yaml`][1] (it was originally`.json`, but changed to `.yaml` in a follow-up PR.  `oc` injects the `0000_50_installer_` prefix when building the release image.

[1]: https://github.com/openshift/installer/pull/4760/files#diff-aa828d811e70962bb290b82c8123d878c2044b831abe0ca56908f61b8300e05cR13

Reply:
@wking as we expand the set of clients that can install a cluster for a topology, the AMI data will need to be elevated out of an “internal” contract and into a “shared” contract.

Reply:
@staebler let’s discuss the above.

Reply:
The contract is to call `openshift-install coreos print-stream-json`. There is no contract for extracting the CoreOS image stream from the release payload directly.

Reply:
The main questions for me:
1. Where is the published schema for the coreos metadata JSON?
2. How should that metadata (conforming to the schema) be made available to users given a release payload image spec (e.g. is there an API call users can make to get the CoreOS metadata for that release, is the the JSON encoded in the image metadata itself, etc.)

Reply:
I would ask the same questions regarding the ImageStream for component images embedded in the payload. In that case the schema is clear (it's an ImageStream which is public versioned schema), but the "how do users get it" is outstanding

Reply:
`oc adm release info -o json $RELEASE_IMAGE_PULLSPEC` will include it.  I'm fine porting that logic to library-go, for other folks who want to vendor it.